src/main/java/eu/dety/burp/joseph/attacks/invalid_curve/ChineseRemainder.java

```diff
@@ -0,0 +1,261 @@
+    private boolean pointCollected = false;
+    private boolean canceled = false;
+
+    private ChineseRemainder() {
+        pointList = new ArrayList<>();
+        // pointList2 = new ArrayList<>();
+    }
+
+    public static ChineseRemainder getInstance() {
+        return instance;
+    }
+
+    public static ChineseRemainder startInstance(ECPublicKey publicKey, ECParameterSpec ecParameters) {
+        instance = new ChineseRemainder();
+        instance.observerList = new ArrayList<>();
+        instance.modulus = BigInteger.ONE;
+        instance.ecParameterSpec = ecParameters;
+        instance.setPublicKey(publicKey);
+        instance.nSquare = ecParameters.getN().pow(2);
+        loggerInstance.log(instance.getClass(), "Single instance CR started.", Logger.LogLevel.DEBUG);
+        return instance;
+    }
+
+    public boolean isEnough() {
+        return modulus.compareTo(nSquare) > 0;
+    }
+
+    public BigInteger calculateModulus(List<Point> moduli) {
+        BigInteger n = BigInteger.ONE;
+        for (Point modulus : moduli) {
+            n = n.multiply(modulus.getOrder());
+        }
+        return n;
+    }
+
+    public BigInteger isqrt(BigInteger integer) {
+        if (integer.compareTo(BigInteger.ZERO) < 0 || integer == null)
+            return null;
+        else if (integer.compareTo(BigInteger.valueOf(2)) < 0)
+            return integer;
+        else {
+            BigInteger floor = isqrt(integer.shiftRight(2)).shiftLeft(1);
+            BigInteger ceiling = floor.add(BigInteger.ONE);
+            if (ceiling.pow(2).compareTo(integer) > 0)
+                return floor;
+            else {
+                return ceiling;
+            }
+        }
+    }
+
+    /**
+     * Set targets public key to compute the private key.
+     *
+     * @param ecPublicKey
+     **/
+    public void setPublicKey(ECPublicKey ecPublicKey) {
+        this.ecPublicKey = ecPublicKey;
+    }
+
+    public synchronized void addPoint(Point point) {
+        if (!pointList.contains(point)) {
+            pointList.add(point);
+            this.pointCollected = true;
+            loggerInstance.log(this.getClass(), "Amount of collected points: " + pointList.size(), Logger.LogLevel.DEBUG);
+        } else
+            return;
+        checkResult(calculateSquaredCR(pointList));
+    }
+
+    // public synchronized void addPoint(Point point) {
+    // if (!pointList.contains(point)) {
+    // pointList.add(point);
+    // // modulus = modulus.multiply(point.getOrder());
+    //
+    // } else if (!pointList2.contains(point)) {
+    // pointList2.add(point);
+    // pointCollected = true;
+    // checkResultThreaded(calculateSquaredCR(pointList2));
+    // }
+    // if (this.isEnough()) {
+    // loggerInstance.log(this.getClass(), "Enough point values collected. ", Logger.LogLevel.DEBUG);
+    // checkResultThreaded(calculateSquaredCR(pointList));
+    // }
+    // }
+
+    public void lastTry() {
```

Comment:
Is it really necessary to perform this computation in threads? why 3 threads?

Reply:
This method is used to calculate the private key with brute force by randomly selecting valid response points and check against the public key. In case of curve P-256, the key is found before this method is called. Despite that, the other two curves could probably deliver only results, if some primes are left out, as my first results with P-256 did, unless I tried to start with some higher primes. So basically, the threads are used to speed up this procedure. I left them at 3, because an user defined amount would need input from the InvalidCurveInfo instance. On the other hand, I think it should be logically refactored to the attacks controller InvalidCurve, whilst it is of use. In fact, I could not find out in this short period of time, why collecting the primes and calculate the chinese remainder do not result in the right key, but skipping some of them did.

Reply:
This is maybe similar what we have observed in our paper, see section 5.3 in https://www.nds.ruhr-uni-bochum.de/media/nds/veroeffentlichungen/2015/09/14/main-full.pdf

I am not sure whether it is necessary to use threads for this.

Reply:
@vunsel Any further update on this?